packages/utils/src/dsn.ts

```diff
@@ -1,127 +1,114 @@
+    host: components.host,
+    port: components.port || '',
+    path: components.path || '',
+    projectId: components.projectId,
+  };
+}
 
-    if (projectId) {
-      const projectMatch = projectId.match(/^\d+/);
-      if (projectMatch) {
-        projectId = projectMatch[0];
+function validateDsn(dsn: Dsn): boolean {
+  if (isDebugBuild()) {
+    const { port, projectId, protocol } = dsn;
+
+    const requiredComponents: ReadonlyArray<keyof DsnComponents> = ['protocol', 'publicKey', 'host', 'projectId'];
+    requiredComponents.forEach(component => {
+      if (!dsn[component]) {
+        throw new SentryError(`Invalid Dsn: ${component} missing`);
       }
+    });
+
+    if (!projectId.match(/^\d+$/)) {
+      throw new SentryError(`Invalid Dsn: Invalid projectId ${projectId}`);
     }
 
-    this._fromComponents({ host, pass, path, projectId, port, protocol: protocol as DsnProtocol, publicKey });
-  }
+    if (!isValidProtocol(protocol)) {
+      throw new SentryError(`Invalid Dsn: Invalid protocol ${protocol}`);
+    }
 
-  /** Maps Dsn components into this instance. */
-  private _fromComponents(components: DsnComponents): void {
-    // TODO this is for backwards compatibility, and can be removed in a future version
-    if ('user' in components && !('publicKey' in components)) {
-      components.publicKey = components.user;
+    if (port && isNaN(parseInt(port, 10))) {
+      throw new SentryError(`Invalid Dsn: Invalid port ${port}`);
     }
-    this.user = components.publicKey || '';
-
-    this.protocol = components.protocol;
-    this.publicKey = components.publicKey || '';
-    this.pass = components.pass || '';
-    this.host = components.host;
-    this.port = components.port || '';
-    this.path = components.path || '';
-    this.projectId = components.projectId;
   }
 
-  /** Validates this Dsn and throws on error. */
-  private _validate(): void {
-    // we only validate in debug mode.  This will fail later anyways.
-    if (isDebugBuild()) {
-      ['protocol', 'publicKey', 'host', 'projectId'].forEach(component => {
-        if (!this[component as keyof DsnComponents]) {
-          throw new SentryError(`${ERROR_MESSAGE}: ${component} missing`);
-        }
-      });
-
-      if (!this.projectId.match(/^\d+$/)) {
-        throw new SentryError(`${ERROR_MESSAGE}: Invalid projectId ${this.projectId}`);
-      }
+  return true;
+}
 
-      if (this.protocol !== 'http' && this.protocol !== 'https') {
-        throw new SentryError(`${ERROR_MESSAGE}: Invalid protocol ${this.protocol}`);
-      }
+/** The Sentry Dsn, identifying a Sentry instance and project. */
+export function makeDsn(from: DsnLike): Dsn {
+  const components = typeof from === 'string' ? dsnFromString(from) : dsnFromComponents(from);
 
-      if (this.port && isNaN(parseInt(this.port, 10))) {
-        throw new SentryError(`${ERROR_MESSAGE}: Invalid port ${this.port}`);
-      }
-    }
-  }
+  validateDsn(components);
+
+  const dsn: Dsn = {
```

Comment:
Not a huge fan of returning objects with `toString` methods directly on it, rather than prototypes. Do we need `toString` at all? Can we instead make it a helper function?

Reply:
I also noticed a test failure, i think the `toString` function in it is the culprit.

Reply:
@mitsuhiko, not a fan either, I didnt want to do too many changes, but since we should be the only ones to consume this, let's try remove it. btw, I already made it a helper fn - the toString that we expose just calls `dsntoString` 😅